### ISSUE TITLE
Can I just disable its setting saving functionality and handle it myself?

### ISSUE THREAD

**Issue by PureDark** [Aragas/Bannerlord.MBOptionScreen#12]
I uses .ini for my mod's settings, I personally like to take control over how settings are saved and loaded.

I only need this lib (or the ModLib I am current using) to provide the UI.

It's like this in my code:

```c#
        [SettingPropertyGroup("Fire Arrow Misc")]
        [SettingProperty("Fire Arrow Allowed Units", 0, 6, "What type of units are allowed to use fire arrows, 0=None, 1=Player, 2=Heroes, 3=Companions, 4=Allies, 5=Enemies, 6=All")]
        [XmlElement]
        public int AllowedUnitType
        {
            get => (int)FireLordConfig.AllowedUnitType;
            set => FireLordConfig.AllowedUnitType = (FireLordConfig.UnitType)value;
        }
```
`FireLordConfig` does all the saving and loading for me. All I need from this lib or ModLib is only the UI part. 
So is it possible to add a API to disable file saving for this lib?

And also 
```c#
[SettingPropertyBool(displayName: "Setting Name", order: 0, requireRestart: false, hintText: "Setting explanation.")]
```
throw this at me:
`Class does not contain a constructor taking 4 parameters ...`
Looks like some v2 annotations doesn't have a constructor taking 4 parameters.
Guess it's not ready for use yet, I am gonna try v1 for the time being.

**Comment by Aragas** [comment]
It's a feature in v2.
You'll need to define your custom [ISettingsFormat](https://github.com/Aragas/Bannerlord.MBOptionScreen/blob/versioning/MBOptionScreen/Settings/Formats/ISettingsFormat.cs) and in [SettingsBase](https://github.com/Aragas/Bannerlord.MBOptionScreen/blob/versioning/MBOptionScreen/Settings/SettingsBase.cs) override [Format](https://github.com/Aragas/Bannerlord.MBOptionScreen/blob/versioning/MBOptionScreen/Settings/SettingsBase.cs#L36) so that they will share with [Extension ](https://github.com/Aragas/Bannerlord.MBOptionScreen/blob/versioning/MBOptionScreen/Settings/Formats/ISettingsFormat.cs#L7) the same value (e.g. json for the json provider)
You can define your custom ISettingFormat in your own assembly, it will be found via reflection. Just make sure to include ``[Version("e1.2.0",  200)]`` attributes like in [json example](https://github.com/Aragas/Bannerlord.MBOptionScreen/blob/versioning/MBOptionScreen/Settings/Formats/JsonSettingsFormat.cs)

**Comment by PureDark** [comment]
Regardless whether I override Format of SettingsBase or not, there will always be a json and a custom extention file generated in 
`Documents\Mount and Blade II Bannerlord\Configs\ModSettings\FireLord`
And the custom extention is only gernerated once and the values doesn't save in it.
···c#
        [Version("e1.1.0", 200)]
        [Version("e1.2.0", 200)]
        internal sealed class FireLordFormat : ISettingsFormat
        {
            public IEnumerable<string> Extensions => new string[] { "firelord" };

            public bool Save(SettingsBase settings, string path)
            {
                if(settings is FireLordSettings fireLordSettings)
                {
                    fireLordSettings.Save();
                }

                return true;
            }

            public SettingsBase Load(SettingsBase settings, string path)
            {
                if (settings is FireLordSettings fireLordSettings)
                {
                    fireLordSettings.Load();
                }
                return settings;
            }
        }
```
I defined this, I want none file be gernerated by this lib, and load and save values only through my own class, but no luck with it yet.

**Comment by Aragas** [comment]
Did you override the SettingsBase.Format?
```csharp
public class Settings : AttributeSettings<Settings >
{
    public override string Format => "firelord";
}
```

**Comment by PureDark** [comment]
I did.
And everytime when I set Format to something other than 'json', or just having a CustomFormat whose Extention contains something other than 'json', I'll get a KeyNotFoundException.

You can easily reproduce it.

**Comment by Aragas** [comment]
Okay, so I tried to reproduce it with the following classes
```csharp
    internal sealed class Test : AttributeSettings<Test>
    {
        public override string Id { get; set; } = "Test";
        public override string ModName => "Test";
        public override string ModuleFolderName => "";
        public override string Format => "firelord";
    }

    [Version("e1.0.0",  200)]
    [Version("e1.0.1",  200)]
    [Version("e1.0.2",  200)]
    [Version("e1.0.3",  200)]
    [Version("e1.0.4",  200)]
    [Version("e1.0.5",  200)]
    [Version("e1.0.6",  200)]
    [Version("e1.0.7",  200)]
    [Version("e1.0.8",  200)]
    [Version("e1.0.9",  200)]
    [Version("e1.0.10", 200)]
    [Version("e1.0.11", 200)]
    [Version("e1.1.0",  200)]
    [Version("e1.2.0",  200)]
    [Version("e1.2.1",  200)]
    [Version("e1.3.0",  200)]
    public class TT : ISettingsFormat
    {
        public IEnumerable<string> Extensions => new string[] { "firelord" };

        public SettingsBase? Load(SettingsBase settings, string path)
        {
            return settings;
        }

        public bool Save(SettingsBase settings, string path)
        {
            return true;
        }
    }
```

It doesn't create a file for me and while I was debugging I did saw that the 'firelord' ISettingsFormat being used by 'firelord' Settings.
Maybe there is something I missed?

**Comment by PureDark** [comment]
That is strange, the stack trace did show that there's something wrong in 
`MBOptionScreen.Settings.SettingsProviderWrapper.RegisterSettings()`

I am gonna create a new mod project to test it, to see whether this problem will arise.

**Comment by PureDark** [comment]
It's now fixed in v2.0.10.
Dunno why I got KeyNotFoundException in earlier version, but I don't really care, I am happy it's fixed.

Love your lib, keep up the good work.
